subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/TaskUpToDateIntegrationTest.groovy

```diff
@@ -50,22 +52,217 @@ class TaskUpToDateIntegrationTest extends AbstractIntegrationSpec {
+                outputFile = project.hasProperty('outputFile') ? file(project.property("outputFile")) : null
             }
         """
+
+        when:
+        succeeds "customTask"
+        then:
+        executedAndNotSkipped ":customTask"
+
+        when:
+        succeeds "customTask", "-PoutputFile=log.txt"
+        then:
+        executedAndNotSkipped ":customTask"
+    }
+
+    @Unroll
+    @Issue("https://github.com/gradle/gradle/issues/3073")
+    def "output files changed from #before to #after marks task #upToDateString"() {
+        buildFile << """
+            class CustomTask extends DefaultTask {
+                @OutputFiles
+                List<Object> outputFiles
+            
+                @TaskAction
+                void doAction() {
+                    outputFiles.each { output ->
+                         def outputFile = output.call()
+                         if (outputFile != null) {
+                            outputFile.text = "task executed"
+                         }
+                    }
+                }
+            }
+            
+            def lazyProperty(String name) {
+                def value = project.findProperty(name)
+                def outputFile = value ? file(value) : null
+                return { -> outputFile }
+            }
+            
+            task customTask(type: CustomTask) {
+                outputFiles = [lazyProperty('output0'), lazyProperty('output1'), lazyProperty('output2')]
+            }
+        """
+
+        when:
+        succeeds customTaskWithOutputs(before)
+        then:
+        executedAndNotSkipped ":customTask"
+
+        when:
+        succeeds customTaskWithOutputs(after)
+        then:
+        if (upToDate) {
+            skipped ":customTask"
+        } else {
+            executedAndNotSkipped ":customTask"
+        }
+
+        where:
+        before                                        | after                          | upToDate
+        [null, 'outputFile']                          | ['outputFile', null]           | true
+        ['outputFile1', null]                         | ['outputFile1', 'outputFile2'] | false
+        ['outputFile1', 'outputFile2', 'outputFile3'] | ['outputFile1', 'outputFile2'] | false
+        ['outputFile1', 'outputFile2']                | ['outputFile2', 'outputFile3'] | false
+        ['outputFile1', 'outputFile2']                | []                             | false
+        ['outputFile1', 'outputFile2']                | ['outputFile1']                | false
+        upToDateString = upToDate ? 'up-to-date' : 'not up-to-date'
+    }
+
+
+    private static String[] customTaskWithOutputs(List<String> outputs) {
+        (["customTask"] + outputs.withIndex().collect { value, idx -> "-Poutput${idx}" + (value ? "=${value}" : '') }) as String[]
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/3073")
+    def "optional input changed from null to non-null marks task not up-to-date"() {
```

Comment:
This should probably be called `"optional input file changed..."`. Should we also add a test for an `@Optional @Input`?

Reply:
We already have that test: https://github.com/gradle/gradle/blob/7b6eafa07ea862c0a2afcfeb234530ac80e1e26b/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskInputPropertiesIntegrationTest.groovy#L405

Reply:
👍